src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -240,149 +240,187 @@ private static ImmutableArray<TextChangeRange> Merge(IReadOnlyList<ImmutableArra
-                    {
-                        // new change is a non-change, just ignore it and move on
-                        newIndex++;
-                        goto nextNewChange;
-                    }
-                    else if (newChange.Span.End < (oldChange.Span.Start + oldDelta))
+            // The sum of characters inserted by old changes minus characters deleted by old changes.
+            // This value must be adjusted whenever characters from an old change are added to `builder`.
+            var oldDelta = 0;
+
+            // In this loop we "zip" together potentially overlapping old and new changes.
+            // It's important that when overlapping changes are found, we don't consume past the end of the overlapping section until the next iteration.
+            // so that we don't miss scenarios where the section after the overlap we found itself overlaps with another change
+            // e.g.:
+            // [ ------- oldChange1 ------ ]
+            // [ -- newChange1 -- ]   [ -- newChange2 --]
+            while (true)
+            {
+                if (oldChange.Span.Length == 0 && oldChange.NewLength == 0)
+                {
+                    if (tryGetNext(oldChanges, ref oldIndex, out oldChange)) continue;
+                    else break;
+                }
+                else if (newChange.Span.Length == 0 && newChange.NewLength == 0)
+                {
+                    if (tryGetNext(newChanges, ref newIndex, out newChange)) continue;
+                    else break;
+                }
+                else if (newChange.Span.End < oldChange.Span.Start + oldDelta)
+                {
+                    // new change is entirely before old change
+                    addAdjustedNewChange(newChange);
+                    if (tryGetNext(newChanges, ref newIndex, out newChange)) continue;
+                    else break;
+                }
+                else if (newChange.Span.Start > oldChange.Span.Start + oldDelta + oldChange.NewLength)
+                {
+                    // new change is entirely after old change
+                    addAndAdjustOldDelta(oldChange);
+                    if (tryGetNext(oldChanges, ref oldIndex, out oldChange)) continue;
+                    else break;
+                }
+                else if (newChange.Span.Start < oldChange.Span.Start + oldDelta)
+                {
+                    // new change starts before old change, but overlaps
+                    // add as much of new change deletion as possible and try again
+                    var newChangeLeadingDeletion = oldChange.Span.Start + oldDelta - newChange.Span.Start;
+                    add(new TextChangeRange(new TextSpan(newChange.Span.Start - oldDelta, newChangeLeadingDeletion), 0));
+                    newChange = new TextChangeRange(new TextSpan(oldChange.Span.Start + oldDelta, newChange.Span.Length - newChangeLeadingDeletion), newChange.NewLength);
```

Comment:
need picture.  this math it not at all clear.

Reply:
IMO, what would likely make this much claerer would be:

1. enumerate the cases where this can happen.
2. check for those cases.
3. for each of those cases, break up either change into the constiuent parts, and then process each part.  i.e. if we have:

```
// [  --          A           --  ]     [  --            B             -- ]
//                  [ --             C                 --]
```

then break this up into:

```
// [  --         A1][A2           --  ]     [  --            B1][B2             -- ]
//                  [ --            C1][ C2][C3              --]
```

(or whatever is actually correct for this case).

Then we can label and properly add/handle each case and see how this impacts the final array.